### ISSUE TITLE
Doubts regarding WMT19 models

### ISSUE THREAD

**Issue by gvskalyan** [pytorch/fairseq#1327]
1. Can you please share the details/configuration of how KENLM laguage model is trained and for cross-entropy scores, is it just model.score(sentence) - to filter out based on cutoff scores.
2. While fine-tuning with domain specific data, are the development and few test sets, added to the original train+BT data or is it only with the former.
3. Does noisy channel work well when the beam size is in the range of 5-10, I don't think so because obviously the beam 10 hypothesis might be better but the scores of LM and channel model might help.
4. How is the length penalty value fixed and is used?
5. Regarding the language models [released](https://github.com/pytorch/fairseq/tree/master/examples/wmt19) since this is a decoder of transformer model is is auto-regressive.
6. Can the above language models be used to score sentences using this 
    `scorer = SequenceScorer(self.task.target_dictionary) #, self.args.softmax_batch)
    hypos = scorer.generate(self.models, sample)`
7. In the paper there are two language models for English in ru-en and de-en directions since the data with which they are trained vary, but why is only one release and is the released model trained on both the data or deduped data of both? Please elaborate.
8. Please provide details/command/procedure to train the language model and filtering/pre-processing for monolingual Newscrawl dataset, or is it same as langid and length filtering.
9. As shown in [this](https://ai.facebook.com/blog/facebook-leads-wmt-translation-competition/) video, how are scores from translation models (are they values from hypothesis score) and language models obtained such as (0.95,0.85,0.82), how are they normalized i.e., to lie less than 1 or so.

Thanks.

@edunov @nng555

**Comment by edunov** [comment]
1.  For kenlm, we didn't do anything special, it was basically running this command: 
lmplz --skip_symbols -o 5 -S 80%   
2. Only the former
3. We found it works best with beam size 50, see Table 2  in this paper for the details: https://arxiv.org/abs/1908.05731
4. Before channel model reranking length penalty is always set to 1, with channel model reranking we tune length penalty on the dev set
5. Not sure, if I understand the question. Language model is implemented as decoder, if you use it to generate text, you'll use it as an autoregressive model. For our purpose, we only need to score candidate sentences, so no autoregression is needed. 
6. Yes, this is how it is done, check eval_lm.py for exact implementation: https://github.com/pytorch/fairseq/blob/master/eval_lm.py
7. Not sure I understand this one either. English language model was trained on English newscrawl, so there is only one. It doesn't depend on the translation direction
8. The example here is pretty good, although it is not exactly the model we trained for the competition: 
https://github.com/pytorch/fairseq/tree/master/examples/language_model  
The difference is most the size and number of layers. The pre-processing for newscrawl is just langid filtering and BPE, and for language models we learned BPE on the newscrawl itself.  
9. We combine those scores with weights that are optimized on the dev set. Please refer to the paper for the details: https://arxiv.org/abs/1908.05731
